src/backend/distributed/executor/intermediate_results.c

```diff
@@ -0,0 +1,699 @@
+							   resultDirectory)));
+	}
+
+	CreatedResultsDirectory = true;
+
+	return resultDirectory;
+}
+
+
+/*
+ * QueryResultFileName returns the file name in which to store
+ * an intermediate result with the given key in the per transaction
+ * result directory.
+ */
+static char *
+QueryResultFileName(const char *resultKey)
+{
+	StringInfo resultFileName = makeStringInfo();
+	const char *resultDirectory = IntermediateResultsDirectory();
+	char *checkChar = (char *) resultKey;
+
+	for (; *checkChar; checkChar++)
+	{
+		if (!((*checkChar >= 'a' && *checkChar <= 'z') ||
+			  (*checkChar >= 'A' && *checkChar <= 'Z') ||
+			  (*checkChar >= '0' && *checkChar <= '9') ||
+			  (*checkChar == '_') || (*checkChar == '-')))
+		{
+			ereport(ERROR, (errcode(ERRCODE_INVALID_NAME),
+							errmsg("result key \"%s\" contains invalid character",
+								   resultKey),
+							errhint("Result keys may only contain letters, numbers, "
+									"underscores and hyphens.")));
+		}
+	}
+
+	appendStringInfo(resultFileName, "%s/%s.data",
+					 resultDirectory, resultKey);
+
+	return resultFileName->data;
+}
+
+
+/*
+ * IntermediateResultsDirectory returns the directory to use for a query result
+ * file with a particular key. The filename includes the user OID, such
+ * that users can never read each other's files.
+ *
+ * When in a distributed transaction block, the directory is of the form:
+ * <user id>_<coordinator node id>_<transaction number>
+ *
+ * Outside of a transaction block the directory is of the form:
+ * <user id>_<process id>
+ *
+ * We choose not to error out outside of a distributed transaction block,
+ * since that would cause issues if IntermediateResultsDirectory is called from the
+ * abort callback. Instead, we allow the user to create process-local
+ * intermediate results, which could be a handy replacement for temporary
+ * tables on secondaries.
+ */
+static char *
+IntermediateResultsDirectory(void)
+{
+	StringInfo resultFileName = makeStringInfo();
+	Oid userId = GetUserId();
+	DistributedTransactionId *transactionId = GetCurrentDistributedTransactionId();
+	int initiatorNodeIdentifier = transactionId->initiatorNodeIdentifier;
+	uint64 transactionNumber = transactionId->transactionNumber;
+
+	if (transactionNumber > 0)
+	{
+		appendStringInfo(resultFileName, "base/" PG_JOB_CACHE_DIR "/%u_%u_%lu",
+						 userId, initiatorNodeIdentifier, transactionNumber);
+	}
+	else
+	{
+		appendStringInfo(resultFileName, "base/" PG_JOB_CACHE_DIR "/%u_%u",
+						 userId, MyProcPid);
+	}
+
+	return resultFileName->data;
+}
+
+
+/*
+ * RemoveIntermediateResultsDirectory removes the intermediate result directory
+ * for the current distributed transaction, if any was created.
+ */
+void
+RemoveIntermediateResultsDirectory(void)
+{
+	if (CreatedResultsDirectory)
+	{
+		StringInfo resultsDirectory = makeStringInfo();
+		appendStringInfoString(resultsDirectory, IntermediateResultsDirectory());
+
+		CitusRemoveDirectory(resultsDirectory);
```

Comment:
`CitusRemoveDirectory()` already checks the existence of the folder, if not returns.

Did you introduce  `CreatedResultsDirectory ` for performance reasons such that `RemoveIntermediateResultsDirectory ` is called for every coordinated tx? 

And, as you've already mentioned in person, if real-time queries do not send `BEGIN`, we'd be in trouble while running the recursive queries.